src/opt/horizontal_fusion_impl.hpp

```diff
@@ -0,0 +1,217 @@
+                        std::unordered_map<instruction_ref, instruction_ref>&,
+                        std::unordered_map<instruction_ref, int>&);
+    void transform();
+    void transform_layers(std::vector<std::vector<instruction_ref>>&,
+                          std::unordered_map<instruction_ref, instruction_ref>&,
+                          int,
+                          std::vector<instruction_ref>&);
+    void
+    transform_output(unsigned,
+                     std::unordered_map<instruction_ref, int>&,
+                     std::unordered_map<instruction_ref, std::vector<std::vector<std::size_t>>>&,
+                     std::unordered_map<instruction_ref, std::vector<int>>&);
+
+    std::vector<instruction_ref> get_instrs(unsigned hash_id)
+    {
+        assert(hash_instrs.find(hash_id) != hash_instrs.end());
+        std::vector<instruction_ref> instrs;
+        for(auto&& point : hash_instrs[hash_id])
+        {
+            assert(point2_instr.find(point) != point2_instr.end());
+            instrs.push_back(point2_instr[point]);
+        }
+        return instrs;
+    }
+    bool compare_inputs(std::vector<instruction_ref>&,
+                        std::vector<instruction_ref>&,
+                        instruction_ref,
+                        int);
+    std::vector<instruction_ref> walk(instruction_ref, std::unordered_map<instruction_ref, bool>&);
+    void concat(std::vector<instruction_ref>&,
+                std::unordered_map<instruction_ref, instruction_ref>&,
+                int);
+    int find_axis(instruction_ref, std::unordered_map<instruction_ref, bool>&);
+    int find_axis(instruction_ref, int dim);
+    int find_axis(instruction_ref, instruction_ref, int);
+    int find_unique_axis(instruction_ref, instruction_ref);
+    std::vector<unsigned> find_cluster(unsigned);
+    bool match_dim(instruction_ref, instruction_ref, int axis);
+    bool is_conv(instruction_ref);
+    bool is_concat(instruction_ref);
+    bool has_unique_output(const std::vector<instruction_ref>&);
+    void remove_redundant_roots(std::vector<instruction_ref>&);
+    void update_hash_tree(unsigned hash_id);
+    int get_channel_axis() { return 1; }
+    int get_conv_output_axis() { return 0; }
+    instruction_ref break_split(int, instruction_ref);
+
+#ifdef MIGRAPHX_DEBUG_OPT
+    void dump_program();
+    void dump_hash_value(hash_value&);
+    void dump_hash_tree();
+#endif
+    private:
+    program* p_program;
+    // Flag an instruction to hash.
+    std::unordered_map<instruction_ref, bool> instr2_hash;
+    // Map an instruction to a hash value pointer.
+    ins2_val instr2_value;
+    std::unordered_map<unsigned, instruction_ref> point2_instr;
+    // Map an encoding to a hash value pointer.
+    std::unordered_map<key_type, hash_value_ptr> encode2_value;
+    // Map an operation name to its encoder function.
+    std::unordered_map<std::string, encoder> op_registry;
+    std::unordered_map<std::string, int> op_flag;
```

Comment:
What does this `int` mean?

Reply:
whether it is gpu::convolution, gpu::conv_bias_relu, or convolution.
W.R.T to the following comment, I currently only hash for convolution, relu and add.  Other operations are assigned a distinct hash id.   Hashing all operations is a task beyond this task, and should be a separated item.   Also global value numbering itself is a big task.

Reply:
Also please specify which cases are not tested.

Reply:
MLIR , TVM and GLOW all realize that they need to take advantage of decades of compiler works.  Rewriting your own pass of compile works is not practical.

Reply:
> whether it is gpu::convolution, gpu::conv_bias_relu, or convolution.

So how does that map to the integers? Does `0` mean `gpu::convolution` and `1` mean `gpu::conv_bias_relu`?

Also this is used before lowering so there isnt a need to support the `gpu::` based operators.

Reply:
means a flag for this operator.

Reply:
1 means it is convolution-related.  We need to know which dimension is channel, where to get filter size and kernel size etc.

Reply:
What does `2` mean? Or maybe use an enum instead to make each of the values clearer? 

But actually I would think it would be better to drop support for target-dependent operators instead. Its never used after lowering, and if we want to support it after lowering, we would want a list of supported operators to be passed in to the constructor, not hard-coded in the class. Instead, we can simplify the code, and just check `ins->name() == "convolution"`.

Reply:
No comment area for next issue.  Why separating encoding out make it easy to write test?

Reply:
Because we can easily write `EXPECT(encode(conv1) == encode(conv2))` without needing to setup `conv1` and `conv2` for horizontal fusion. We can also test when the encoding should not be equal, which is more difficult with layers of value numbering and horizontal fusion layered on top. For example, we can show that `encode(relu) != encode(relu)` in a graph where horizontal fusion would be applied.

Reply:
Also separating this out, we also make the code easier to follow. It is not entirely clear to me what code is for encoding or value numbering or for horizontal fusion. It would be nice to have clearly defined interfaces for those with tests to understand what the inputs and outputs expects.